src/basicfuns.jl

```diff
@@ -232,19 +232,18 @@ That is, `r` is overwritten with `exp.(x)`, normalized to sum to 1.
 
 See the [Wikipedia entry](https://en.wikipedia.org/wiki/Softmax_function)
 """
-function softmax!(r::AbstractArray{R}, x::AbstractArray{T}) where {R<:AbstractFloat,T<:Real}
+function softmax!(r::AbstractArray{R}, x::AbstractArray{T}) where {R<:Real, T<:Real}
     n = length(x)
     length(r) == n || throw(DimensionMismatch("Inconsistent array lengths."))
     u = maximum(x)
-    s = 0.
+    s = zero(R)
     @inbounds for i = 1:n
         s += (r[i] = exp(x[i] - u))
     end
-    invs = convert(R, inv(s))
```

Comment:
I guess it is faster to compute the inverse only once but it would be interesting to check if this is actually true. In any case, I don't think it is needed to convert the inverse since `s` is initialized based on `eltype(r)` now.